### PR TITLE
Special-case suggestions for null pointers constness cast

### LINKED ISSUE #13361: Suggest `ptr::null_mut::<T>()` instead of `ptr::null::<T>() as *mut T`
### What it does

Use the `null_mut()` directly instead of first creating a read-only `null()` and casting it

### Advantage

- shorter easier to read code
- avoid accidental later mis-casting

### Drawbacks

none

### Example

```rust
let a = ptr::null::<T>() as *mut T;
let b = ptr::null_mut::<T>() as *const T;
```

Could be written as:

```rust
let a = ptr::null_mut::<T>();
let b = ptr::null::<T>();
```

### PR DESCRIPTION
**Pull request by samueltardieu**
This implements the suggestions from #13361. It fits into the existing `ptr_cast_constness` lint, as this is a specialized version. However,

1. I have not modified the lint MSRV, so the documentation for this lint will still show that it applies only from Rust 1.72.0. This is true in the general case, but the lint for null pointers will trigger even before this version as `null()` and `null_mut()` were already present in Rust 1.0 and there is no reason not to apply this lint. I guess this is only a minor documentation issue that can be ignored.
2. I have not covered the `core::ptr::null::<T>().cast_mut()` (could be made into `core::ptr::null_mut::<T>()`) and `cotr::ptr::null_mut::<T>().cast_const()` (could be made into `core::ptr::null::<T>()`) cases. Should they be covered? If they should, here or in a separate PR?

changelog: [`ptr_cast_constness`]: special-case suggestions for null pointers constness cast

Fix #13361

### CODE REVIEW

**Comment by y21** [comment]
For the second point, it sounds reasonable to me to lint on that too. Doesn't need to be done in this PR and could be a followup if you want to do that also, up to you. The linked issue also doesn't mention it and this is already an improvement on its own.

File: clippy_lints/src/casts/ptr_cast_constness.rs [hunk 1]
```diff
@@ -26,20 +26,54 @@ pub(super) fn check<'tcx>(
...
+            let mut app = Applicability::MachineApplicable;
+            let sugg = Sugg::hir_with_applicability(cx, cast_expr, "_", &mut app).to_string();
+            if cx.tcx.is_diagnostic_item(sym::ptr_null, method_defid) {
+                span_lint_and_sugg(
+                    cx,
+                    PTR_CAST_CONSTNESS,
+                    expr.span,
+                    "`as` casting to make a const null pointer into a mutable null pointer",
+                    "use `null_mut()` directly instead",
+                    sugg.replace("null", "null_mut"),
```
[the remark this answers is not in the document]
Comment (samueltardieu) [reply clippy_lints/src/casts/ptr_cast_constness.rs]: Done

**Comment by samueltardieu** [comment]
> For the second point, it sounds reasonable to me to lint on that too as part of the same lint (although the description would then need to be adapted a fair bit since it's drifting away from the original intent of linting on `as` casts specifically). Doesn't need to be done in this PR and could be a followup if you want to do that also, up to you. The linked issue also doesn't mention it and this is already an improvement on its own.

I've added it in a second commit.

File: clippy_lints/src/casts/ptr_cast_constness.rs [hunk 2]
```diff
@@ -26,20 +27,87 @@ pub(super) fn check<'tcx>(
...
+                format!("{}.cast_{constness}()", sugg.maybe_par()),
+                Applicability::MachineApplicable,
+            );
+        }
+    }
+}
+
+pub(super) fn check_null_ptr_cast_method(cx: &LateContext<'_>, expr: &Expr<'_>) {
+    if let ExprKind::MethodCall(method, cast_expr, [], _) = expr.kind
+        && matches!(method.ident.as_str(), "cast_mut" | "cast_const")
```
Comment (y21) [inline clippy_lints/src/casts/ptr_cast_constness.rs:86]: Can you move this check further down to be after the hir checks and store the method ident in a variable somewhere so the interner only needs to be locked once? Currently there are three calls to Symbol::as_str

Comment (y21) [reply clippy_lints/src/casts/ptr_cast_constness.rs:86]: Also this `matches` and the ptr/ptr_mut diagnostic item checks can be merged together, would simplify the code a bit
```suggestion
        && let mutable = match (cx.tcx.get_diagnostic_name(defid), method.ident.as_str()) {
            (Some(sym::ptr_null), "cast_mut") => true,
            (Some(sym::ptr_null_mut), "cast_const") => false,
            _ => return,
        };
```

File: clippy_lints/src/casts/mod.rs [hunk 3]
```diff
@@ -410,19 +410,23 @@ declare_clippy_lint! {
     /// ### Why is this bad?
```
[the remark this answers is not in the document]
Comment (samueltardieu) [reply clippy_lints/src/casts/mod.rs:410]: Done. I've also numbered the pointers in the example to make it clearer which solution corresponds to which problem.

File: clippy_lints/src/casts/ptr_cast_constness.rs (hunk 2 above)
Comment (samueltardieu) [reply clippy_lints/src/casts/ptr_cast_constness.rs:86]: I moved the `let` statements into the `if` (which simplifies further the previous `let Some(…) else { return; };`). Indeed, using `get_diagnostic_name()` makes it easier than `is_diagnostic_name()` here.

**Review by y21 — APPROVED** [review]
This looks good to me as is. I think having specific lint messages are fine :+1:

**Comment by y21** [comment]
Thank you! @bors r+

**Pull request merged**